### PR TITLE
Secure message

### PR DESCRIPTION
**Pull request by IsaacYangSLA**
### Description

Add CellCipher for secure message encryption/decryption
Add SessionKeyManager to handle key exchange and management

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [x] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/fuel/f3/cellnet/cell_cipher.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            algorithm=hashes.SHA256(),
+        )
+        remote_pub_key = remote_cert.public_key()
+        key_enc = remote_pub_key.encrypt(
+            session_key,
+            padding.OAEP(mgf=padding.MGF1(algorithm=hashes.SHA256()), algorithm=hashes.SHA256(), label=None),
+        )
+        self.key_hash_dict[get_hash(session_key)[-HASH_LENGTH:]] = session_key
+        key_response = {"key_enc": bytes_to_b64str(key_enc), "signature": bytes_to_b64str(signature)}
+        return key_response
```
Comment (nvidianz) [inline nvflare/fuel/f3/cellnet/cell_cipher.py:63]: key_response should be bytes, not dictionary.

File: nvflare/fuel/f3/cellnet/cell_cipher.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+class SessionKeyUnavailable(Exception):
+    pass
+
+
+class SessionKeyManager:
+    def __init__(self):
+        self.key_hash_dict = dict()
+
+    def key_request(self, remote_cert, local_cert, local_pri_key):
+        session_key = os.urandom(KEY_LENGTH)
```
Comment (nvidianz) [inline nvflare/fuel/f3/cellnet/cell_cipher.py:47]: Should use a key derivation algorithm to generate key. Random number could yield weak keys.

File: nvflare/fuel/f3/cellnet/cell_cipher.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            ),
+            algorithm=hashes.SHA256(),
+        )
+        remote_pub_key = remote_cert.public_key()
+        key_enc = remote_pub_key.encrypt(
+            session_key,
+            padding.OAEP(mgf=padding.MGF1(algorithm=hashes.SHA256()), algorithm=hashes.SHA256(), label=None),
+        )
+        self.key_hash_dict[get_hash(session_key)[-HASH_LENGTH:]] = session_key
+        key_response = {"key_enc": bytes_to_b64str(key_enc), "signature": bytes_to_b64str(signature)}
```
Comment (nvidianz) [inline nvflare/fuel/f3/cellnet/cell_cipher.py:62]: Should encrypt the key first, then sign the whole thing

File: nvflare/fuel/f3/cellnet/cell_cipher.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    hash.update(value)
+    return hash.finalize()
+
+
+class SessionKeyUnavailable(Exception):
+    pass
+
+
+class SessionKeyManager:
+    def __init__(self):
```
Comment (nvidianz) [inline nvflare/fuel/f3/cellnet/cell_cipher.py:43]: To validate cert, CA certs and intermediate certs are needed. SessionManager constructor needs those information.

File: nvflare/fuel/f3/cellnet/cell_cipher.py (hunk 2 above)
Comment (IsaacYangSLA) [reply nvflare/fuel/f3/cellnet/cell_cipher.py:49]: This is recommended by cryptography package.

File: nvflare/fuel/f3/cellnet/cell_cipher.py (hunk 4 above)
Comment (IsaacYangSLA) [reply nvflare/fuel/f3/cellnet/cell_cipher.py:45]: We only need root CA cert.  There is no intermediate cert.  I will add the root ca cert as argument in the init.

File: nvflare/fuel/f3/cellnet/cell_cipher.py (hunk 3 above)
Comment (IsaacYangSLA) [reply nvflare/fuel/f3/cellnet/cell_cipher.py]: Do you mean sign the clear and encrypted key?

File: nvflare/fuel/f3/cellnet/cell_cipher.py (hunk 1 above)
Comment (IsaacYangSLA) [reply nvflare/fuel/f3/cellnet/cell_cipher.py:65]: Pushed a new commit.

**Review by nvidianz — APPROVED** [review]

**Pull request merged**